rust/lance-index/src/scalar/expression.rs

```diff
@@ -54,6 +58,166 @@ pub struct IndexedExpression {
+    }
+
+    fn visit_is_bool(&self, column: &str, value: bool) -> Option<IndexedExpression> {
+        Some(IndexedExpression::index_query(
+            column.to_string(),
+            Arc::new(SargableQuery::Equals(ScalarValue::Boolean(Some(value)))),
+        ))
+    }
+
+    fn visit_is_null(&self, column: &str) -> Option<IndexedExpression> {
+        Some(IndexedExpression::index_query(
+            column.to_string(),
+            Arc::new(SargableQuery::IsNull()),
+        ))
+    }
+
+    fn visit_comparison(
+        &self,
+        column: &str,
+        value: ScalarValue,
+        op: &Operator,
+    ) -> Option<IndexedExpression> {
+        let query = match op {
+            Operator::Lt => SargableQuery::Range(Bound::Unbounded, Bound::Excluded(value)),
+            Operator::LtEq => SargableQuery::Range(Bound::Unbounded, Bound::Included(value)),
+            Operator::Gt => SargableQuery::Range(Bound::Excluded(value), Bound::Unbounded),
+            Operator::GtEq => SargableQuery::Range(Bound::Included(value), Bound::Unbounded),
+            Operator::Eq => SargableQuery::Equals(value),
+            // This will be negated by the caller
+            Operator::NotEq => SargableQuery::Equals(value),
+            _ => unreachable!(),
+        };
+        Some(IndexedExpression::index_query(
+            column.to_string(),
+            Arc::new(query),
+        ))
+    }
+
+    fn visit_scalar_function(
+        &self,
+        _: &str,
+        _: &DataType,
+        _: &ScalarFunctionDefinition,
+        _: &[Expr],
+    ) -> Option<IndexedExpression> {
+        None
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct TagQueryParser {}
+
+impl ScalarQueryParser for TagQueryParser {
+    fn visit_between(&self, _: &str, _: ScalarValue, _: ScalarValue) -> Option<IndexedExpression> {
+        None
+    }
+
+    fn visit_in_list(&self, _: &str, _: Vec<ScalarValue>) -> Option<IndexedExpression> {
+        None
+    }
+
+    fn visit_is_bool(&self, _: &str, _: bool) -> Option<IndexedExpression> {
+        None
+    }
+
+    fn visit_is_null(&self, _: &str) -> Option<IndexedExpression> {
+        None
+    }
+
+    fn visit_comparison(&self, _: &str, _: ScalarValue, _: &Operator) -> Option<IndexedExpression> {
+        None
+    }
+
+    fn visit_scalar_function(
+        &self,
+        column: &str,
+        data_type: &DataType,
+        func: &ScalarFunctionDefinition,
+        args: &[Expr],
+    ) -> Option<IndexedExpression> {
+        if args.len() != 2 {
+            return None;
+        }
+        let tag_list = maybe_scalar(&args[1], data_type)?;
+        if let ScalarValue::List(list_arr) = tag_list {
+            let list_values = list_arr.values();
+            let mut scalars = Vec::with_capacity(list_values.len());
+            for idx in 0..list_values.len() {
+                scalars.push(ScalarValue::try_from_array(list_values.as_ref(), idx).ok()?);
+            }
+            if func.name() == "array_has_all" {
```

Comment:
IMO we should also handle `array_has`. (`array_contains` is an alias for this, I believe). https://datafusion.apache.org/user-guide/sql/scalar_functions.html#array-contains

Reply:
Good point, we can map `array_has` to `array_has_any`